google/api_core/retry_streaming.py

```diff
@@ -0,0 +1,220 @@
+        # remove seen items from request on each attempt
+        new_request = modify_request(request, seen_items)
+        new_generator = target(new_request)
+        for item in new_generator:
+            yield item
+            seen_items.append(item)
+
+    retry_wrapped = Retry(is_stream=True)(attempt_with_modified_request, target, request, [])
+    ```
+
+    2. Wrap the retry generator
+        Alternatively, you can wrap the retryable generator itself before
+        passing it to the end-user to add a filter on the stream. For
+        example, you can keep track of the items that were successfully yielded
+        in previous retry attempts, and only yield new items when the
+        new attempt surpasses the previous ones:
+
+        ``
+        def retryable_with_filter(target):
+            stream_idx = 0
+            # reset stream_idx when the stream is retried
+            def on_error(e):
+                nonlocal stream_idx
+                stream_idx = 0
+            # build retryable
+            retryable_gen = Retry(is_stream=True, on_error=on_error, ...)(target)
+            # keep track of what has been yielded out of filter
+            yielded_items = []
+            for item in retryable_gen:
+                if stream_idx >= len(yielded_items):
+                    yield item
+                    yielded_items.append(item)
+                elif item != previous_stream[stream_idx]:
+                    raise ValueError("Stream differs from last attempt")"
+                stream_idx += 1
+
+        filter_retry_wrapped = retryable_with_filter(target)
+        ```
+"""
+
+from typing import (
+    Callable,
+    Optional,
+    List,
+    Tuple,
+    Iterable,
+    Generator,
+    TypeVar,
+    Any,
+    cast,
+)
+
+import logging
+import time
+from functools import partial
+
+from google.api_core import exceptions
+
+_Y = TypeVar("_Y")  # yielded values
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _build_timeout_error(
+    exc_list: List[Exception], is_timeout: bool, timeout_val: float
+) -> Tuple[Exception, Optional[Exception]]:
+    """
+    Default exception_factory implementation. Builds an exception after the retry fails
+
+    Args:
+      - exc_list (list[Exception]): list of exceptions that occurred during the retry
+      - is_timeout (bool): whether the failure is due to the timeout value being exceeded,
+          or due to a non-retryable exception
+      - timeout_val (float): the original timeout value for the retry, for use in the exception message
+
+    Returns:
+      - tuple[Exception, Exception|None]: a tuple of the exception to be raised, and the cause exception if any
+    """
+    src_exc = exc_list[-1] if exc_list else None
+    if is_timeout:
+        return (
+            exceptions.RetryError(
+                "Timeout of {:.1f}s exceeded".format(timeout_val),
+                src_exc,
+            ),
+            src_exc,
+        )
+    else:
+        return exc_list[-1], None
```

Comment:
I think the problem was with mypy, since src_exc can be None but we want to always return an exception.

I made some changes to fall back to a default error if no exceptions are passed though. Let me know what you think